File: extra_1/app/src/main/java/com/example/extra_1/api/NetworkSingleton.java
package com.example.extra_1.api;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class NetworkSingleton {
    private static volatile WeatherApi weatherApi;

    private NetworkSingleton() {}

    public static WeatherApi getWeatherApi() {
        if (weatherApi == null) {
            synchronized (NetworkSingleton.class) {
                if (weatherApi == null) {
                    Retrofit retrofit = new Retrofit.Builder()
                            .baseUrl("https://api.openweathermap.org/")
                            .addConverterFactory(GsonConverterFactory.create())
                            .build();
                    weatherApi = retrofit.create(WeatherApi.class);
                }
            }
        }
        return weatherApi;
    }
}

File: extra1/app/src/main/java/com/example/extra1/api/NetworkSingleton.kt
package com.example.extra1.api

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object NetworkSingleton {
    private val retrofit: Retrofit by lazy {
        Retrofit.Builder()
            .baseUrl("https://api.openweathermap.org/")
            .addConverterFactory(GsonConverterFactory.create())
            .build()
    }

    val weatherApi: WeatherApi by lazy { retrofit.create(WeatherApi::class.java) }
}
